p2p/transport/webrtc/stream_read.go

```diff
@@ -0,0 +1,102 @@
+package libp2pwebrtc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+	"github.com/pion/datachannel"
+)
+
+// Read from the underlying datachannel. This also
+// process sctp control messages such as DCEP, which is
+// handled internally by pion, and stream closure which
+// is signaled by `Read` on the datachannel returning
+// io.EOF.
+func (s *webRTCStream) Read(b []byte) (int, error) {
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	var (
+		readErr error
+		read    int
+	)
+	for read == 0 && readErr == nil {
+		if s.isClosed() {
+			return 0, io.ErrClosedPipe
+		}
+		read, readErr = s.readMessage(b)
+	}
+	return read, readErr
+}
+
+func (s *webRTCStream) readMessage(b []byte) (int, error) {
+	read := copy(b, s.readBuffer)
+	s.readBuffer = s.readBuffer[read:]
+	remaining := len(s.readBuffer)
+
+	if remaining == 0 && !s.stateHandler.AllowRead() {
+		log.Debug("[2] stream has no more data to read")
+		return read, io.EOF
+	}
+
+	if read > 0 {
+		return read, nil
+	}
+
+	// read from datachannel
+	var msg pb.Message
+	err := s.readMessageFromDataChannel(&msg)
+	if err != nil {
+		// This case occurs when the remote node goes away
+		// without writing a FIN message
+		if errors.Is(err, io.EOF) {
+			s.Reset()
+			return read, io.ErrClosedPipe
+		}
+		return read, err
+	}
+
+	// append incoming data to read readBuffer
+	if s.stateHandler.AllowRead() && msg.Message != nil {
+		s.readBuffer = append(s.readBuffer, msg.GetMessage()...)
+	}
+
+	// process any flags on the message
+	if msg.Flag != nil {
+		s.processIncomingFlag(msg.GetFlag())
+	}
+	return read, nil
+}
+
+func (s *webRTCStream) readMessageFromDataChannel(msg *pb.Message) error {
+	s.readerMux.Lock()
```

Comment:
Not sure what the anti pattern would be here? A mutex is used for a single shared resource while a semaphore is used for multiple shared resources? Feel free to point me to reading material as I have not a storng opinion on this, but I do not get why a mutex use would be wrong here? And using a channel just because you can (wrapped in a semaphore or directly) also seems a bit odd here.

So going to close this, but feel free to open it with more info in case you have a strong opinion on this matter. I'm ok either way.